can't run executable with spaces in path on windows

I rarely deal with windows, so maybe I'm missing something. But I feel like I've tried most things that could reasonably be expected to work. This is all on the just-released v1.3.2.

I have an executable in `C:\Users\Public\with spaces\node.exe`, and I would like nodemon to run it. This is what happened:

This would be nice, but I don't expect it to work:
    
    > node nodemon\bin\nodemon.js --exec "C:\Users\Public\with spaces\node.exe" -- -c console.log(1)
    21 Jan 16:13:01 - [nodemon] v1.3.2
    21 Jan 16:13:01 - [nodemon] to restart at any time, enter `rs`
    21 Jan 16:13:01 - [nodemon] watching: *.*
    21 Jan 16:13:01 - [nodemon] starting `C:\Users\Public\with spaces\node.exe -c console.log(1)`
    'C:\Users\Public\with' is not recognized as an internal or external command, operable program or batch file.
    
how about some double-quotes?
    
    >node bin\nodemon.js --exec "\"C:\Users\Public\with spaces\node.exe\"" -- -c console.log(1)
    21 Jan 16:14:22 - [nodemon] v1.3.2
    21 Jan 16:14:22 - [nodemon] to restart at any time, enter `rs`
    21 Jan 16:14:22 - [nodemon] watching: *.*
    21 Jan 16:14:22 - [nodemon] starting `"C:\Users\Public\with spaces\node.exe" -c console.log(1)`
    '\"C:\Users\Public\with spaces\node.exe\"' is not recognized as an internal or external command, operable program or batch file.
  
double-double-quotes?
    
    >node bin\nodemon.js --exec "\"\"C:\Users\Public\with spaces\node.exe\"\"" -- -c console.log(1)
    21 Jan 16:15:06 - [nodemon] v1.3.2
    21 Jan 16:15:06 - [nodemon] to restart at any time, enter `rs`
    21 Jan 16:15:06 - [nodemon] watching: *.*
    21 Jan 16:15:06 - [nodemon] starting `""C:\Users\Public\with spaces\node.exe"" -c console.log(1)`
    The network name cannot be found.
    
well that's a different error, but that's probably not a good thing?
    
maybe I'll pass it as one big --exec option?
    
    >node bin\nodemon.js --exec "\"C:\Users\Public\with spaces\node.exe\" -c console.log(1)"
    21 Jan 16:17:19 - [nodemon] v1.3.2
    21 Jan 16:17:19 - [nodemon] to restart at any time, enter `rs`
    21 Jan 16:17:19 - [nodemon] watching: *.*
    21 Jan 16:17:19 - [nodemon] starting `"C:\Users\Public\with spaces\node.exe" -c console.log(1)`
    '\"C:\Users\Public\with spaces\node.exe\"' is not recognized as an internal or external command,
    operable program or batch file.
    
Maybe cmd wants me to wrap the whole thing in another set of double quotes?
    
    >node bin\nodemon.js --exec "\"\"C:\Users\Public\with spaces\node.exe\" -c console.log(1)\""
    21 Jan 16:18:03 - [nodemon] v1.3.2
    21 Jan 16:18:03 - [nodemon] to restart at any time, enter `rs`
    21 Jan 16:18:03 - [nodemon] watching: *.*
    21 Jan 16:18:03 - [nodemon] starting `""C:\Users\Public\with spaces\node.exe" -c console.log(1)"`
    The network name cannot be found.

You're probably sick of me filing bugs on special characters. Feel free to close it if you don't want to reproduce / fix, I'm just filing it to make sure the issue is known.

I no longer mind if you fix this bug or not (and I probably won't try to reproduce on later versions), at this point I'm just going to use a minimal fork which doesn't pass anything through `sh` or `cmd`. I mean no disrespect by forking - I thank you for all your hard work on `nodemon`, but for my use case it's just not worth diagnosing and working around these endless escaping issues.

That's news to me (weird that cmd /? doesn't say anything about it).

I gave it a go, and the results are underwhelming:

    C:>node bin\nodemon.js --exec "C:\Users\Public\with^ spaces\node.exe" -- -e console.log(1)
    22 Jan 14:02:44 - [nodemon] v1.3.2
    22 Jan 14:02:44 - [nodemon] to restart at any time, enter `rs`
    22 Jan 14:02:44 - [nodemon] watching: *.*
    22 Jan 14:02:44 - [nodemon] starting `C:\Users\Public\with^ spaces\node.exe -e console.log(1)`
    
    module.js:340
        throw err;
              ^
    Error: Cannot find module 'C:\Users\Public\with spaces\spaces\node.exe'
        at Function.Module._resolveFilename (module.js:338:15)
        at Function.Module._load (module.js:280:25)
        at Function.Module.runMain (module.js:497:10)
        at startup (node.js:119:16)
        at node.js:906:3
    22 Jan 14:02:44 - [nodemon] app crashed - waiting for file changes before starting...


It _seems_ to be starting node, and then something's going wildly wrong given the module path it seems to be trying to load. It's hard to tell whether this is a `node` or `cmd` bug (my money is on `cmd`, or just "windows" ;)). Regardless of whose bug it is, it's definitely triggered by the `cmd` indirection:

    C:>"C:\Users\Public\with spaces\node.exe" -e console.log(1)
    1
    
    C:>cmd /c "C:\Users\Public\with^ spaces\node.exe" -e console.log(1)
    
    module.js:340
        throw err;
              ^
    Error: Cannot find module 'C:\Users\Public\with spaces\spaces\node.exe'
        at Function.Module._resolveFilename (module.js:338:15)
        at Function.Module._load (module.js:280:25)
        at Function.Module.runMain (module.js:497:10)
        at startup (node.js:119:16)
        at node.js:906:3

Have you actually tried to run nodemon (on windows) with an executable path with spaces in it? I really think this deserves to be part of some integration test cases (not just unit tests), given how problematic it seems to be...

--- a/lib/config/exec.js
+++ b/lib/config/exec.js
@@ -17,6 +17,7 @@ function parseExecutable(str) {
   var unescapeNextTime = false;
   var open = false;
   var c = null;
+  var isescape = false;
 
   for (; i < length; i++) {
     c = str[i];
@@ -27,11 +28,23 @@ function parseExecutable(str) {
 
     unescapeNextTime = escaped;
 
-    if (c === '\\') {
+    if (c === '\\' && !utils.isWindows) {
+      isescape = true;
+    } else if (c === '^' && utils.isWindows) {
+      isescape = true;
+    } else {
+      isescape = false;
+    }
+
+    if (isescape) {
       escaped = true;
       // peek next character
       if (str[i+1] === ' ') { // space we skip
         inspace = true;
+      }
+
+      if (utils.isWindows) {
+        continue; // chomp the carat
       }
     }
 

--- a/lib/monitor/run.js
+++ b/lib/monitor/run.js
@@ -55,7 +55,21 @@ function run(options) {
     return arg;
   });
 
-  args = [cmd.executable].concat(args).join(' ').trim();
+  var executable = cmd.executable;
+
+  // special logic for windows, as spaces in the paths need the path fragment
+  // quoted, so it reads: c:\"Program Files"\nodejs\node.exe
+  if (utils.isWindows && executable.indexOf(' ') !== -1) {
+    executable = executable.split('\\').map(function (part) {
+      if (part.indexOf(' ') !== -1) {
+        return '"' + part.replace(/"/g, '\\"') + '"';
+      }
+      return part;
+    }).join('\\');
+  }
+
+  args = [executable].concat(args).join(' ').trim();
+
   var spawnArgs = [sh, [shFlag, args]];
 
   if (nodeMajor === 0 && nodeMinor < 8) {

